lib/api/core/validation/types/date.js

```diff
@@ -0,0 +1,211 @@
+    strict_week_date: date => moment.utc(date, 'gggg-\\Www-e', true),
+    week_date_time: date => moment.utc(date, 'gggg-\\Www-e\\THH:mm:ss.SSSZZ', false),
+    strict_week_date_time: date => moment.utc(date, 'gggg-\\Www-e\\THH:mm:ss.SSSZZ', true),
+    week_date_time_no_millis: date => moment.utc(date, 'gggg-\\Www-e\\THH:mm:ssZZ', false),
+    strict_week_date_time_no_millis: date => moment.utc(date, 'gggg-\\Www-e\\THH:mm:ssZZ', true),
+    weekyear: date => moment.utc(date, 'gggg', false),
+    strict_weekyear: date => moment.utc(date, 'gggg', true),
+    weekyear_week: date => moment.utc(date, 'ggggww', false),
+    strict_weekyear_week: date => moment.utc(date, 'ggggww', true),
+    weekyear_week_day: date => moment.utc(date, 'ggggwwe', false),
+    strict_weekyear_week_day: date => moment.utc(date, 'ggggwwe', true),
+    year: date => moment.utc(date, 'YYYY', false),
+    strict_year: date => moment.utc(date, 'YYYY', true),
+    year_month: date => moment.utc(date, 'YYYYMM', false),
+    strict_year_month: date => moment.utc(date, 'YYYYMM', true),
+    year_month_day: date => moment.utc(date, 'YYYYMMDD', false),
+    strict_year_month_day: date => moment.utc(date, 'YYYYMMDD', true)
+  };
+
+/**
+ * @constructor
+ */
+function DateType () {
+  this.typeName = 'date';
+  this.allowChildren = false;
+  this.allowedTypeOptions = ['range', 'formats'];
+}
+
+DateType.prototype = new BaseConstructor();
+
+/**
+ * @param {TypeOptions} typeOptions
+ * @param {*} fieldValue
+ * @param {string[]} errorMessages
+ */
+DateType.prototype.validate = function (typeOptions, fieldValue, errorMessages) {
+  var
+    i,
+    momentDate,
+    valid = false;
+
+  for (i = 0; i < typeOptions.formats.length; i++) {
+    momentDate = formatMap[typeOptions.formats[i]](fieldValue);
+    if (momentDate.isValid()) {
+      valid = true;
+      break;
+    }
+  }
+
+  if (!valid) {
+    errorMessages.push('The date format is not valid.');
+    return false;
+  }
+
+  if (typeOptions.hasOwnProperty('range')) {
+    if (typeOptions.range.min && momentDate.isBefore(typeOptions.range.min)) {
+      errorMessages.push('The provided date is before the defined minimum.');
+      return false;
+    }
+
+    if (typeOptions.range.max && typeOptions.range.max.isBefore(momentDate)) {
+      errorMessages.push('The provided date is after the defined maximum.');
+      return false;
+    }
+  }
+
+  return true;
+};
+
+/**
+ * @param {TypeOptions} typeOptions
+ * @return {boolean|TypeOptions}
+ * @throws InternalError
+ */
+DateType.prototype.validateFieldSpecification = function (typeOptions) {
+  var
+    result = true,
+    min,
+    max;
+
+  if (typeOptions.hasOwnProperty('formats')) {
+    if (!Array.isArray(typeOptions.formats) || typeOptions.formats.length === 0) {
+      return false;
+    }
+
+    typeOptions.formats.forEach(format => {
+      if (Object.keys(formatMap).indexOf(format) === -1) {
+        result = false;
+        return false;
```

Comment:
same as in validation/index.js. Looks like an invalid attempt to escape forEach.